### ISSUE TITLE
Basic example Raspberry Pi Python without Bridge

### ISSUE THREAD

**Issue by CesarJG** [Sensirion/python-i2c-scd#1]
Hi, 

first of all many thanks for the effort made developing the drivers for this CO2 sensor.

At https://sensirion.github.io/python-i2c-scd/quickstart.html#sensorbridge-example there is a basic example using the SensorBridge. Could you possible provide a python example for users without the SensorBridge? (Sensor directly connected to Raspberry Pi GPIO pins). 

I am trying to figure out which sensor to buy (SCD30 or SCD41). Would this library also work for SCD30?

Many thanks in advance and best regards.

**Comment by MBjoern** [comment]
Hi there

Thank you for using our sensors and libraries.

We have lots of examples and dirvers available for the SCD30 and SCD40 sensor: See [here](https://github.com/sensirion?q=scd)

For Raspberry Pi via GPIOs, we have drivers working out of the box for the SCD4x [here](https://github.com/Sensirion/raspberry-pi-i2c-scd4x) and for the SCD30 (requiring minor tweaks by copying the linux i2c implementation into the root directory of the driver) [here](https://github.com/Sensirion/embedded-scd/releases/download/2.1.2/scd30-2.1.2.zip)

Have fun testing and prototyping and let us know what Use Case you're developing.

**Comment by CesarJG** [comment]
Hi, 

thanks for your response but I was specifically asking for a python example without the SensorBridge. Something like here: https://sensirion.github.io/python-i2c-scd/quickstart.html#sensorbridge-example .I think that would be very helpful for non experienced users without the SensorBridge.

Regards.

**Comment by CesarJG** [comment]
Hi, 

thanks for your response but I think all the provided examples are written in c. I was asking for a python example like at https://sensirion.github.io/python-i2c-scd/quickstart.html#sensorbridge-example . I think that would be a perfect addition for non experienced users without the SensorBridge.

Best regards.

**Comment by psachs** [comment]
Hi,

This python driver is for SCD41 only. Currently we don't have a public available python driver for SCD30. You can however also use this driver with the I2C bridge. To do so the sensor needs to be connected to the standard i2c pins (GPIO 2 => SDA; GPIO3 => SLC) of the raspberry pi and you need to install our python i2c driver:
https://pypi.org/project/sensirion-i2c-driver/



```python
from sensirion_i2c_driver import LinuxI2cTransceiver, I2cConnection
from sensirion_i2c_scd import Scd4xI2cDevice
 
with LinuxI2cTransceiver('/dev/i2c-1') as i2c_transceiver:
    i2c_connection = I2cConnection(i2c_transceiver)    
    scd41 = Scd4xI2cDevice(i2c_connection)

    # start periodic measurement in high power mode
    scd41.start_periodic_measurement()

    # Measure every 5 seconds
    while True:
        time.sleep(5)
        co2, temperature, humidity = scd41.read_measurement()
        # use default formatting for printing output:
        print("{}, {}, {}".format(co2, temperature, humidity))
        # custom printing of attributes:
        print("{:d} ppm CO2, {:0.2f} °C ({} ticks), {:0.1f} %RH ({} ticks)".format(
            co2.co2,
            temperature.degrees_celsius, temperature.ticks,
            humidity.percent_rh, humidity.ticks))
    scd41.stop_periodic_measurement()
```

**Comment by CesarJG** [comment]
Hi again, 

I think that is what I was looking for. I think I will go for the SCD41 :)

Many thanks for your time and keep doing such a good job.

**Comment by CesarJG** [comment]
Hi, 

Today arrived my SCD41 Evaluation Board and I would like to inform you that the above example seems not to work. I become:

python3 co2-test.py 
Traceback (most recent call last):
  File "co2-test.py", line 10, in <module>
    scd41.start_periodic_measurement()
  File "/usr/local/lib/python3.7/dist-packages/sensirion_i2c_scd/scd4x/device.py", line 56, in start_periodic_measurement
    result = self.execute(Scd4xI2cCmdStartPeriodicMeasurement())
  File "/usr/local/lib/python3.7/dist-packages/sensirion_i2c_driver/device.py", line 61, in execute
    return self._connection.execute(self._slave_address, command)
  File "/usr/local/lib/python3.7/dist-packages/sensirion_i2c_driver/connection.py", line 116, in execute
    return self._interpret_response(command, response)
  File "/usr/local/lib/python3.7/dist-packages/sensirion_i2c_driver/connection.py", line 205, in _interpret_response
    raise response
sensirion_i2c_driver.errors.I2cTransceiveError: I2C transceive failed: [Errno 121] Remote I/O error


I am somewhat confused since "i2cdetect -y 1" shows the address of the sensor (among others that I already had):

     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- 0a -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: 50 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- 62 -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- 76 --  

The command "i2cdetect -l|sort" outputs the following: 

i2c-1	i2c       	bcm2835 I2C adapter             	I2C adapter


The funny thing is that with the tutorial from https://github.com/Sensirion/raspberry-pi-i2c-scd4x written in c everything worked out of the box, with the sensor printing the data without any problem. Maybe there is something wrong with the python library? Please let me know if you would like more info. For now I will try to make a workaround by launching the c program and parse its output from inside a python script.

Best regards.

**Comment by MBjoern** [comment]
Hi there

I was able to reproduce the same error when trying the above script (with some more imports though).
I also checked the i2cdetect and `i2cget -y 1 0x62 0` to check if the sensor is responsive and made sure the c code works properly...

In the end i got it to work after power cycling the SCD4x (unplug vdd and replug and try to run the script again). Did you try that?

Let me know how it goes.

Best regards
Björn

**Comment by rbuffat** [comment]
I ran into the same issue (Running the example code in a `python:3.9` based Docker container on a raspberry pi 4).

Currently, it looks as the following commands help to mitigate the issue. But I have yet to confirm if this is works reliably. 
```
...
scd41 = Scd4xI2cDevice(i2c_connection)
scd41.stop_periodic_measurement()
time.sleep(1)
scd41.reinit()
time.sleep(5)
# start periodic measurement in high power mode
scd41.start_periodic_measurement()
...
```

It is based on https://github.com/Sensirion/raspberry-pi-i2c-scd4x/blob/master/scd4x_i2c_example_usage.c#L48-L51
Interestingly,  `scd41.wake_up()` will result in the same error when used before `scd41.stop_periodic_measurement()` contrary to the raspberry pi C sample code.

**Comment by psachs** [comment]
I re-open this issue so we can better track the underlying problem.
The Python driver should behave the same as the raspberry PI C driver as they both connect to /dev/i2c-1 to communicate with the sensor.

Since we access a file-like device we have to ensure that only once process is communicating with it at the same time. Else we might get some I/O errors.

**Comment by psachs** [comment]
I tested all scripts and input several times on a Raspberry Pi with an SCD41 connected to i2c-1 and could not reproduce any of the errors.

I therefore assume it is not a problem with the Python driver.

**Comment by MycroHarvest** [comment]
To get @psachs 's code to work I also had to add this line to the imports

from sensirion_i2c_scd import Scd4xI2cDevice

After installing the scd package with 

pip install sensirion-i2c-scd

**Comment by psachs** [comment]
@MycroHarvest  thanks for the note.
Of course you are correct. I will edit my response to fix the missing import in the example.

**Comment by bocephasj** [comment]
I am having an issue with using the 'scd4x.wake_up()' command. I am trying to put sensor in sleep(scd4x.power_down() ) mode and then using the (low power) scd4x.measure_single_shot() [taking reading every 6hrs] and I get errno 5 - I/O error when I send the wake_up() command. Is there something I am doing wrong?
All other commands work fine.  --Doing this on Raspberry Pi Zero 2W, Python 3.11--

**Comment by psachs** [comment]
the correct measurement process is as follows. If you execute it in this order everything should work. Please not that during power down, every command will wake up the sensor, so make sure you don't communicate with the sensor while on sleep. The wake up command will only be successful if the sensor is in sleep mode.

```python
scd4x.wake_up()
scd4x.measure_single_shot()
scd4x.power_down()
```

**Comment by bocephasj** [comment]
That is exactly how I did it and I get: 
'sensirion_i2c_driver.errors.I2cTransceiveError: I2C transceive failed: [Error 5] Input/output error'
this ONLY happens when using the  'scd4x.wake_up()' command, I have even gone so far as run all commands individually and it only happens when executing the 'scd4x.wake_up()' command. Thank you for responding to this issue.

**Comment by bocephasj** [comment]
This issue does reoccur with any command after entering sleep mode (scd4x.power_down())

**Comment by psachs** [comment]
We will have a look into this issue. Can we please create a new issue for this, as this one is unrelated and already closed.

**Comment by psachs** [comment]
Hi @bocephasj 

I can reproduce your issue. The reason for this is that the sensors responds with a NACK which is thrown as a I2cTransceiveError (resp. I/O error).

This error is expected on wake up. If you catch the error and ignore it the sensor should be in idle mode and a single shot measurement can be performed.

You should be able to do something like that for single shot measurements:

```python
import time
from sensirion_i2c_driver import LinuxI2cTransceiver, I2cConnection
from sensirion_i2c_scd import Scd4xI2cDevice
from sensirion_i2c_driver.errors import I2cTransceiveError

with LinuxI2cTransceiver("/dev/i2c-1") as i2c_transceiver:
    # Create SCD4x device
    scd4x = Scd4xI2cDevice(I2cConnection(i2c_transceiver))

    # Ensure we are start in idle mode
    try:
        scd4x.reinit()
    except I2cTransceiveError:
        pass

    print("scd4x Serial Number: {}".format(scd4x.read_serial_number()))

    scd4x.power_down()

    # Measure every minute for 10 minutes
    for _ in range(10):
        time.sleep(60)
        try:
            scd4x.wake_up()
        except I2cTransceiveError:
            pass
        scd4x.measure_single_shot()
        co2, temperature, humidity = scd4x.read_measurement()
        # use default formatting for printing output:
        print("{}, {}, {}".format(co2, temperature, humidity))
        scd4x.power_down()
```

**Comment by bocephasj** [comment]
This is the new error:

 File "....../sensirion_i2c_driver/connection.py", line 205, in _interpret_response
    raise response
sensirion_i2c_driver.errors.I2cTransceiveError: I2C transceive failed: [Errno 5] Input/output error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "......../test2_scd4x.py", line 22, in <module>
    except I2cTransceiveError:
NameError: name 'I2cTransceiveError' is not defined

In the section:
    # Ensure we are start in idle mode
    try:
      - scd4x.reinit() - (what ever command I put here errors)
    except I2cTransceiveError:
        pass
Could the problem be in 'response_types' script 'looking' for a response of some sort that is not being handled by the error handling routine?

**Comment by psachs** [comment]
Are you on the latest version of the driver? I tested this on Raspberry Pi with a new installation of the sensirion-i2c-scd driver.
The error shows you that I2cTransceiveError is not defined, which looks like we are not using the same version of the sensirion-i2c-driver package (https://github.com/Sensirion/python-i2c-driver).
You need version 1.0.0 of sensirion-i2c-driver for it to work.

Alternative you can also replace it with an IOError which is part of the python standard installation
```
except IOError:
    pass
```

**Comment by bocephasj** [comment]
My apologies: I missed the line 'from sensirion_i2c_driver.errors import I2cTransceiveError
'
It works! Now I will try to run from cron. Thank you very much for the quick response and the right answers!
